### PR TITLE
Handle new data correctly and extend functionality of MMM posterior predictive methods

### LINKED ISSUE #472: scaling control vars
Hi all - I think it’s easy for an unscaled control to go unnoticed. I think it’d be beneficial to either:
1. provide the option to scale control vars in the same way channels are in the `DelayedSaturatedMMM` class, or
2. raise a warning to the user if controls appear to not be scaled

### PR DESCRIPTION
**Pull request by wd60622**
## Description

Redo of the support for new mmm data

This brings in the feedback about 

- having the `include_last_observations` for all predict methods with `X_pred`
- Support for original scale of target with the `original_scale` keyword that can be passed to all predict methods

## Related Issue

- [ ] Closes #472 #450 #400 #396 
- [ ] Related to #444 #392 

## Checklist

- [x] Checked that [the pre-commit linting/style checks pass](https://docs.pymc.io/en/latest/contributing/python_style.html)
- [x] Included tests that prove the fix is effective or that the new feature works
- [x] Added necessary documentation (docstrings and/or example notebooks)
- [ ] If you are a pro: each commit corresponds to a [relevant logical change](https://wiki.openstack.org/wiki/GitCommitMessages#Structural_split_of_changes)

## Modules affected

- [x] MMM
- [ ] CLV

## Type of change

- [x] New feature / enhancement
- [ ] Bug fix
- [ ] Documentation
- [ ] Maintenance
- [ ] Other (please specify):

----
📚 Documentation preview 📚: https://pymc-marketing--482.org.readthedocs.build/en/482/

### CODE REVIEW

File: tests/mmm/test_delayed_saturated_mmm.py
Comment (wd60622) [inline tests/mmm/test_delayed_saturated_mmm.py:642]: @juanitorduz 
I can still break these down a bit more
I'm trying to get a handful of the edge cases via the parametrize tests

Are you thinking one for each predict method?

Comment (juanitorduz) [reply tests/mmm/test_delayed_saturated_mmm.py:642]: Yes, ideally test should be atomic (test one thing) and have all assert statements at the end. Not a strong opinion, more.of a suggestion :)

Comment (wd60622) [reply tests/mmm/test_delayed_saturated_mmm.py:642]: Fair enough. I'm pulling out the date ranges and separating the tests

**Comment by wd60622** [comment]
Introducing he new `predict_posterior` call on mmm model

Screenshot 2024-01-11 at 20 41 28

File: pymc_marketing/mmm/delayed_saturated_mmm.py
Comment (ricardoV94) [inline pymc_marketing/mmm/delayed_saturated_mmm.py:87]: I think y is a better good name. likelihood is about the kind of variable it is. That would be like renaming betas to "prior"

File: pymc_marketing/mmm/delayed_saturated_mmm.py
Comment (ricardoV94) [inline pymc_marketing/mmm/delayed_saturated_mmm.py:911]: This kind of complexity/gotchas go away if you make the transforms part of the model itself (but for that you need to write them in PyTensor)

That's why I pushed for ir before. Out of the scope of this PR

File: pymc_marketing/mmm/delayed_saturated_mmm.py
Comment (ricardoV94) [inline pymc_marketing/mmm/delayed_saturated_mmm.py:895]: You need to define join otherwise calling this method twice will ignore the second run

See https://github.com/pymc-devs/pymc-experimental/pull/249

Comment (ricardoV94) [reply pymc_marketing/mmm/delayed_saturated_mmm.py:895]: This applies to prior and fit as well, but maybe it's a good time as any to fix it as it's a pretty serious one.

I had a PR on this but haven't come back to it

File: pymc_marketing/mmm/delayed_saturated_mmm.py
Comment (wd60622) [reply pymc_marketing/mmm/delayed_saturated_mmm.py:87]: I change in the model definition then!

File: pymc_marketing/mmm/delayed_saturated_mmm.py
Comment (wd60622) [reply pymc_marketing/mmm/delayed_saturated_mmm.py:911]: I know we've had the discussion when talking about #407. Maybe it is easier to start with the target as a pytensor transformation. 
Any change back to original scale (a few spots) would have to be changed

File: pymc_marketing/mmm/delayed_saturated_mmm.py
Comment (wd60622) [reply pymc_marketing/mmm/delayed_saturated_mmm.py:895]: Sounds good. This is in the method override. Shall I add to the model_builder method as well?

Comment (ricardoV94) [reply pymc_marketing/mmm/delayed_saturated_mmm.py:895]: Yup. Feel free to reuse as much as you can from the PR of the pymc-experimental ModelBuilder (thinking about tests as well). Just add the author as a co-author if you do that.

Comment (wd60622) [reply pymc_marketing/mmm/delayed_saturated_mmm.py]: Sounds good. I will dive into it more later

Comment (wd60622) [reply pymc_marketing/mmm/delayed_saturated_mmm.py]: I've just added a check that the idata will be the first sample from two method calls to keep it simple. Hope that's what you expected

File: pymc_marketing/mmm/delayed_saturated_mmm.py
Comment (juanitorduz) [inline pymc_marketing/mmm/delayed_saturated_mmm.py:869]: Remove last (repeated point)

File: pymc_marketing/mmm/delayed_saturated_mmm.py
Comment (juanitorduz) [inline pymc_marketing/mmm/delayed_saturated_mmm.py:874]: Maybe add a reason why this is relevant. I mean, a  comment that this is relevant for the adstock computation.

File: pymc_marketing/mmm/delayed_saturated_mmm.py
Comment (juanitorduz) [inline pymc_marketing/mmm/delayed_saturated_mmm.py:887]: Shall we sort by date just in case?

File: pymc_marketing/mmm/delayed_saturated_mmm.py
Comment (juanitorduz) [inline pymc_marketing/mmm/delayed_saturated_mmm.py:861]: Maybe call this `sample_posterior_predictive_kwargs` (or something more informative?) To indicate where these additional parameters will be used.

File: pymc_marketing/mmm/delayed_saturated_mmm.py
Comment (juanitorduz) [inline pymc_marketing/mmm/delayed_saturated_mmm.py:922]: I understand this because I wrote it but it might be confusing. So you could write a little helper function to wrap this logic and add a unit-test (just a suggestion)

File: pymc_marketing/mmm/delayed_saturated_mmm.py
Comment (juanitorduz) [reply pymc_marketing/mmm/delayed_saturated_mmm.py:861]: Or as below a better description: "Additional arguments to pass to sample_posterior_predictive method"

**Review by juanitorduz — CHANGES_REQUESTED** [review]
This looks super nice! Thanks! I just added some comments about the docs and other minor things.

**Comment by juanitorduz** [comment]
@wd60622 After we merge this one, we need to add (or create) an example to the docs (let's create an issue).  This one will be a game-changer feature 🚀

File: pymc_marketing/mmm/delayed_saturated_mmm.py
Comment (ricardoV94) [inline pymc_marketing/mmm/delayed_saturated_mmm.py:890]: I wonder if we should reset the data changes after sampling

File: tests/mmm/test_delayed_saturated_mmm.py
Comment (ricardoV94) [inline tests/mmm/test_delayed_saturated_mmm.py:707]: I would like to test something about the output values, not just their shapes

File: tests/mmm/test_delayed_saturated_mmm.py
Comment (ricardoV94) [inline tests/mmm/test_delayed_saturated_mmm.py:647]: What's the difference between the `predict` and `sample_posterior_predictive` MMM methods?

File: tests/model_builder/test_model_builder.py
Comment (ricardoV94) [inline tests/model_builder/test_model_builder.py:300]: Shouldn't it override the first with the second when `extend_idata` is True? 

Also we should parametrize this test to test both `extend_idata` cases

Comment (ricardoV94) [reply tests/model_builder/test_model_builder.py:300]: Also I think these changes also apply to `fit`?

File: tests/mmm/test_delayed_saturated_mmm.py
Comment (wd60622) [reply tests/mmm/test_delayed_saturated_mmm.py:647]: predict is the mean of the sample_posterior_predictive by obs

**Comment by wd60622** [comment]
Thank you guys for the review! I will have some time tomorrow to make some changes

File: pymc_marketing/mmm/delayed_saturated_mmm.py
Comment (wd60622) [reply pymc_marketing/mmm/delayed_saturated_mmm.py:869]: done

File: pymc_marketing/mmm/delayed_saturated_mmm.py
Comment (wd60622) [reply pymc_marketing/mmm/delayed_saturated_mmm.py:874]: done

File: pymc_marketing/mmm/delayed_saturated_mmm.py
Comment (wd60622) [reply pymc_marketing/mmm/delayed_saturated_mmm.py:922]: Will do!

File: pymc_marketing/mmm/delayed_saturated_mmm.py
Comment (wd60622) [reply pymc_marketing/mmm/delayed_saturated_mmm.py:890]: I'm hear this as setting the MutableData back to the X and y that it is fit on. i.e. self.X and self.y
Is that correct?
I think that would be good to preserve

File: tests/model_builder/test_model_builder.py
Comment (wd60622) [reply tests/model_builder/test_model_builder.py:300]: This is the consequent from join="left"
I would think if you call the method again, you'd want to replace or be notified that it is being replaced. 

I think this is out of this PR's scope though\

Comment (ricardoV94) [reply tests/model_builder/test_model_builder.py:300]: We should join right, otherwise extend_inferencedata=True makes little sense?

Comment (wd60622) [reply tests/model_builder/test_model_builder.py:300]: That's what I meant to say. That's how I've change it. join="right"

File: pymc_marketing/mmm/delayed_saturated_mmm.py
Comment (wd60622) [reply pymc_marketing/mmm/delayed_saturated_mmm.py:895]: done

File: pymc_marketing/mmm/delayed_saturated_mmm.py
Comment (wd60622) [reply pymc_marketing/mmm/delayed_saturated_mmm.py:898]: Added a `_reset_model_data` method to set to the original and it is called after the sample_posterior_predictive call

File: tests/mmm/test_delayed_saturated_mmm.py
Comment (wd60622) [reply tests/mmm/test_delayed_saturated_mmm.py:716]: added some checks based on the original scale and the original y for context

File: pymc_marketing/mmm/delayed_saturated_mmm.py
Comment (wd60622) [reply pymc_marketing/mmm/delayed_saturated_mmm.py]: done. This isn't changed for the original model_builder methods however. Would you want that to change in this PR?

Comment (juanitorduz) [reply pymc_marketing/mmm/delayed_saturated_mmm.py]: Ok I think we can leave it as it is in the docstrings for now. Let's wait for feedback!

**Review by juanitorduz — APPROVED** [review]
This is amazing @wd60622 ! From my side, I just wanna push this out and get feedback. My sleep deprivation (👶 ) might make me miss some details so if @ricardoV94 is happy I  suggest we merge and release 0.4 🚀

File: tests/mmm/test_delayed_saturated_mmm.py
Comment (ricardoV94) [inline tests/mmm/test_delayed_saturated_mmm.py:731]: Why this change?

File: tests/model_builder/test_model_builder.py
Comment (ricardoV94) [reply tests/model_builder/test_model_builder.py:300]: This test is still weird.

I expect the assert to be `assert sample == second_sample`.

The fact it is not failing tells me that assert is too premissive. I guess inferencedata equality is just based on groups or something. We need to check some actual values

**Review by ricardoV94 — CHANGES_REQUESTED** [review]

File: pymc_marketing/mmm/delayed_saturated_mmm.py
Comment (ricardoV94) [reply pymc_marketing/mmm/delayed_saturated_mmm.py:898]: On a second thought I see that all sampling methods need X as input, so I think we should revert this change. Otherwise we should reset in all the other methods. Just in this one doesn't make sense.

Apologies, I find the ModelBuilder API pretty confusing.

File: pymc_marketing/model_builder.py
Comment (wd60622) [inline pymc_marketing/model_builder.py:585]: @ricardoV94 the join kwarg was added. Maybe you are looking at old commit

File: pymc_marketing/model_builder.py
Comment (wd60622) [inline pymc_marketing/model_builder.py:619]: Same here. Maybe I missed a spot, so let me know

File: tests/mmm/test_delayed_saturated_mmm.py
Comment (wd60622) [reply tests/mmm/test_delayed_saturated_mmm.py:731]: Not sure. I will change back

File: tests/model_builder/test_model_builder.py
Comment (wd60622) [reply tests/model_builder/test_model_builder.py:300]: Okay. I will take a look later

File: pymc_marketing/model_builder.py
Comment (ricardoV94) [reply pymc_marketing/model_builder.py:619]: I saw those. What are you referring to? I mentioned `fit` will still override the previous result, which is not something that needs to be fixed in this PR.

The test has the wrong assert because it's checking that the idata matches the first call (not the second). The test is also insufficient, because it doesn't raise (I think you didn't change it when you switched from "left" to "right"

File: pymc_marketing/mmm/delayed_saturated_mmm.py
Comment (ricardoV94) [inline pymc_marketing/mmm/delayed_saturated_mmm.py:863]: Default missing? Docstrings say it defaults to False, All other methods seem to have a default.

I am not sure what the default should be, probably True?

Comment (wd60622) [reply pymc_marketing/mmm/delayed_saturated_mmm.py]: I've changed all but the sample_prior_predictive (from model_builder) to default to True. Changed in the docstring as well

File: pymc_marketing/model_builder.py
Comment (wd60622) [reply pymc_marketing/model_builder.py]: I modified the tests in order to check for the second sample stays. Please take a look

**Comment by wd60622** [comment]
Hi @ricardoV94. Thanks for the feedback. I just went through all of it and made some changes. 

For the xarray tests, I changed to used check for the second value and am using xr.testing module now for more robust checks.

**Comment by juanitorduz** [comment]
LGTM 🚀 ! Let's wait for @ricardoV94 's feedback :)

**Comment by ricardoV94** [comment]
@wd60622 everything looks kosher, but there is still this open discussion: https://github.com/pymc-labs/pymc-marketing/pull/482#discussion_r1457076712

Given how all sampling/fitting methods require x/y anyway I changed my mind and I don't think the `_reset_data` call is needed after all? The less magic probably the better

**Comment by wd60622** [comment]
> @wd60622 everything looks kosher, but there is still this open discussion: [#482 (comment)](https://github.com/pymc-labs/pymc-marketing/pull/482#discussion_r1457076712)
> 
> Given how all sampling/fitting methods require x/y anyway I changed my mind and I don't think the `_reset_data` call is needed after all? The less magic probably the better|

Perfect. I've changed that back so that the data will not be reset

File: tests/mmm/test_delayed_saturated_mmm.py
Comment (ricardoV94) [inline tests/mmm/test_delayed_saturated_mmm.py:659]: This check should now also be removed. Having a test suggests it's a feature when it's just what it is. This could be confusing for someone refactoring the code in the future because it suggests it's something that needs to be satisfied.

**Comment by juanitorduz** [comment]
Are we ready to merge? 🤞

**Comment by juanitorduz** [comment]
It's green! 💚

**Review by ricardoV94 — APPROVED** [review]

**Comment by ricardoV94** [comment]
Merged, any notebooks that need updating?

**Comment by juanitorduz** [comment]
> Merged, any notebooks that need updating?

I ran the notebook in my first review and it worked perfectly. I can check tomorrow (@wd60622 do we need an update?).

**Comment by wd60622** [comment]
Firstly, thank you both for the reviews! 

> > Merged, any notebooks that need updating?
> 
> I ran the notebook in my first review and it worked perfectly. I can check tomorrow (@wd60622 do we need an update?).

I haven't touched the notebooks. I would like to add an example in docstring / notebook with https://github.com/pymc-labs/pymc-marketing/issues/494
I will check get to that on the weekend

**Pull request merged**
